src/iso9660.rs

```diff
@@ -0,0 +1,364 @@
+        boot_id: parse_iso9660_string(buf, 32)?,
+    })
+}
+
+/// Parses primary descriptor at cursor.
+fn parse_primary_descriptor(buf: &mut Bytes) -> Result<PrimaryVolumeDescriptor> {
+    verify_descriptor_header(buf).context("parsing primary descriptor")?;
+    let system_id = parse_iso9660_string(eat(buf, 1), 32).context("parsing system id")?;
+    let volume_id = parse_iso9660_string(buf, 32).context("parsing volume id")?;
+    let root = match get_next_directory_record(eat(buf, 156 - 72))? {
+        Some(DirectoryRecord::Directory(d)) => d,
+        _ => bail!("failed to parse root directory record from primary descriptor"),
+    };
+    Ok(PrimaryVolumeDescriptor {
+        system_id,
+        volume_id,
+        root,
+    })
+}
+
+/// Verifies descriptor header at cursor.
+fn verify_descriptor_header(buf: &mut Bytes) -> Result<()> {
+    const VOLUME_DESCRIPTOR_ID: &[u8] = b"CD001";
+    const VOLUME_DESCRIPTOR_VERSION: u8 = 1;
+
+    let id = buf.copy_to_bytes(5);
+    if id != VOLUME_DESCRIPTOR_ID {
+        bail!("unknown descriptor ID: {:?}", id);
+    }
+
+    let version = buf.get_u8();
+    if version != VOLUME_DESCRIPTOR_VERSION {
+        bail!("unknown descriptor version: {}", version);
+    }
+
+    Ok(())
+}
+
+pub struct IsoFsIterator {
+    dir: Bytes,
+}
+
+impl Iterator for IsoFsIterator {
+    type Item = Result<DirectoryRecord>;
+    fn next(&mut self) -> Option<Self::Item> {
+        get_next_directory_record(&mut self.dir)
+            .context("reading next record")
+            .transpose()
+    }
+}
+
+/// Reads the directory record at cursor and advances to the next one.
+fn get_next_directory_record(buf: &mut Bytes) -> Result<Option<DirectoryRecord>> {
+    let len = buf.get_u8() as usize;
+    if len == 0 {
+        return Ok(None);
+    }
+
+    let address = eat(buf, 1).get_u32_le();
+    let length = eat(buf, 4).get_u32_le();
+    let flags = eat(buf, 25 - 14).get_u8();
+    let name_length = eat(buf, 32 - 26).get_u8() as usize;
+    let name = parse_iso9660_path(buf, name_length).context("parsing record name")?;
+
+    // advance to next record
+    eat(buf, len - (33 + name_length));
+
+    if flags & 2 > 0 {
+        Ok(Some(DirectoryRecord::Directory(Directory {
+            name,
+            address,
+            length,
+        })))
+    } else {
+        Ok(Some(DirectoryRecord::File(File {
+            name,
+            address,
+            length,
+        })))
+    }
+}
+
+/// Reads a directory record path. This is similar to a regular ISO9660 string, but supports '\0'
+/// to mean current directory, and '\1' for the parent directory.
+fn parse_iso9660_path(buf: &mut Bytes, len: usize) -> Result<String> {
+    if len == 1 && (buf[0] == 0 || buf[0] == 1) {
+        let c = buf.get_u8();
+        if c == 0 {
+            Ok(".".into())
+        } else {
+            Ok("..".into())
+        }
+    } else {
+        parse_iso9660_string(buf, len)
```

Comment:
ISO 9660 doesn't allow `/` characters here (they're not in the D-string set), but we are, which potentially allows directory traversals.